agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,315 @@
+                                         nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  nr_php_wraprec_hashmap_t* hashmap;
+
+  if (0 == log2_num_buckets) {
+    /*
+     * Encode the default value in one place: namely, here.
+     */
+    log2_num_buckets = 8;
+  } else if (log2_num_buckets > 24) {
+    /*
+     * Basic sanity check: it's extremely unlikely that we'll ever need a
+     * hashmap for the user function wraprecs that has more than 2^24 buckets.
+     */
+    log2_num_buckets = 24;
+  }
+
+  hashmap = (nr_php_wraprec_hashmap_t*)nr_malloc(sizeof(nr_php_wraprec_hashmap_t));
+  hashmap->dtor_func = dtor_fn;
+  hashmap->log2_num_buckets = log2_num_buckets;
+  hashmap->buckets = (nr_wraprecs_bucket_t**)nr_calloc(
+      (1 << log2_num_buckets), sizeof(nr_wraprecs_bucket_t*));
+  hashmap->elements = 0;
+
+  return hashmap;
+}
+
+nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_buckets(size_t buckets, nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+
+  size_t actual_buckets;
+
+  if (0 == buckets) {
+    actual_buckets = 0;
+  } else {
+    /*
+     * Calculate ceil(log2(buckets)), since that's what
+     * nr_hashmap_create_internal requires.
+     */
+
+    actual_buckets = 1;
+    while (((size_t)(1 << actual_buckets)) < buckets) {
+      actual_buckets += 1;
+    }
+  }
+
+  return nr_php_wraprec_hashmap_create_internal(actual_buckets, dtor_fn);
+}
+
+static void nr_destroy_wraprecs_bucket(nr_wraprecs_bucket_t** bucket_ptr,
+                               nr_php_wraprec_hashmap_dtor_fn_t dtor_func) {
+  nr_wraprecs_bucket_t* bucket = *bucket_ptr;
+
+  if (dtor_func) {
+    (dtor_func)(bucket->wraprec);
+  }
+  nr_realfree((void**)bucket_ptr);
+}
+
+nr_php_wraprec_hashmap_stats_t nr_php_wraprec_hashmap_destroy(nr_php_wraprec_hashmap_t** hashmap_ptr) {
+  nr_php_wraprec_hashmap_stats_t stats = {};
+  size_t count;
+  nr_php_wraprec_hashmap_t* hashmap;
+  size_t i;
+  size_t collisions;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return stats;
+  }
+  hashmap = *hashmap_ptr;
+
+  stats.elements = hashmap->elements;
+  stats.collisions_min = stats.elements;
+
+  count = nr_count_buckets(hashmap);
+  for (i = 0; i < count; i++) {
+    nr_wraprecs_bucket_t* bucket = hashmap->buckets[i];
+    collisions = 0;
+
+    if (bucket) {
+      stats.buckets_used++;
+    }
+
+    while (bucket) {
+      nr_wraprecs_bucket_t* next = bucket->next;
+
+      collisions++;
+
+      nr_destroy_wraprecs_bucket(&bucket, hashmap->dtor_func);
+      bucket = next;
+    }
+
+    if (0 != collisions && collisions < stats.collisions_min) {
+      stats.collisions_min = collisions;
+    }
+    if (collisions > stats.collisions_max) {
+      stats.collisions_max = collisions;
+    }
+    if (collisions > 1) {
+      stats.buckets_with_collisions++;
+      stats.collisions_mean += collisions;
+    }
+  }
+
+  if (0 != stats.buckets_with_collisions) {
+    stats.collisions_mean /= stats.buckets_with_collisions;
```

Comment:
Why do we not want to divide by buckets_used instead?

Reply:
`collisions_mean` is only updated when number of elements in bucket is greater than 1. So is `buckets_with_collisions`. `buckets_used` is total number of buckets, even the ones without collisions (i.e. with number of elements in bucket equal to 1). Therefore dividing by `buckets_used` would not give the mean of collisions for buckets with collisions. It would give the mean of collisions including buckets without collisions.